google/resource_spanner_instance.go

```diff
@@ -0,0 +1,330 @@
+package google
+
+import (
+	"fmt"
+	"log"
+	"net/http"
+	"regexp"
+	"strings"
+
+	"github.com/hashicorp/terraform/helper/resource"
+	"github.com/hashicorp/terraform/helper/schema"
+
+	"google.golang.org/api/googleapi"
+	"google.golang.org/api/spanner/v1"
+)
+
+func resourceSpannerInstance() *schema.Resource {
+	return &schema.Resource{
+		Create: resourceSpannerInstanceCreate,
+		Read:   resourceSpannerInstanceRead,
+		Update: resourceSpannerInstanceUpdate,
+		Delete: resourceSpannerInstanceDelete,
+		Importer: &schema.ResourceImporter{
+			State: resourceSpannerInstanceImportState,
+		},
+
+		Schema: map[string]*schema.Schema{
+
+			"config": &schema.Schema{
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
+
+			"name": &schema.Schema{
+				Type:     schema.TypeString,
+				Optional: true,
+				Computed: true,
+				ForceNew: true,
+				ValidateFunc: func(v interface{}, k string) (ws []string, errors []error) {
+					value := v.(string)
+
+					if len(value) < 6 && len(value) > 30 {
+						errors = append(errors, fmt.Errorf(
+							"%q must be between 6 and 30 characters in length", k))
+					}
+					if !regexp.MustCompile("^[a-z0-9-]+$").MatchString(value) {
+						errors = append(errors, fmt.Errorf(
+							"%q can only contain lowercase letters, numbers and hyphens", k))
+					}
+					if !regexp.MustCompile("^[a-z]").MatchString(value) {
+						errors = append(errors, fmt.Errorf(
+							"%q must start with a letter", k))
+					}
+					if !regexp.MustCompile("[a-z0-9]$").MatchString(value) {
+						errors = append(errors, fmt.Errorf(
+							"%q must end with a number or a letter", k))
+					}
+					return
+				},
+			},
+
+			"display_name": &schema.Schema{
+				Type:     schema.TypeString,
+				Required: true,
+				ValidateFunc: func(v interface{}, k string) (ws []string, errors []error) {
+					value := v.(string)
+
+					if len(value) < 4 && len(value) > 30 {
+						errors = append(errors, fmt.Errorf(
+							"%q must be between 4 and 30 characters in length", k))
+					}
+					return
+				},
+			},
+
+			"num_nodes": &schema.Schema{
+				Type:     schema.TypeInt,
+				Optional: true,
+				Default:  1,
+			},
+
+			"labels": {
+				Type:     schema.TypeMap,
+				Optional: true,
+				ForceNew: true,
```

Comment:
The patch API support labels updates. Remove this `ForceNew` and add support for it in the update method below.